### PR TITLE
Gather data on profile customization and tab settings to help us improve defaults

### LINKED ISSUE #3855: Add ETL events to better understand custom profiles globals & tab profile settings used
# Description of the new feature/enhancement

ETL events for understanding the following:
1. CustomKeybindings
2. CustomDefaultProfile
3. Use of acrylic in profiles
4. Use of font face in profiles
5. Tab profile details to understand shells used

# Proposed technical implementation details (optional)

### PR DESCRIPTION
**Pull request by miniksa**
## Summary of the Pull Request
This will collect some user choices related to profiles and tab settings to help us understand if and how we should change the in-built defaults.

## PR Checklist
* [x] Closes #3855
* [x] I work here.
* [x] Manual test only.
* [x] Meh, no doc update.
* [x] Am core contributor.

## Detailed Description of the Pull Request / Additional comments
The following data is collected with examples of the types of questions we intend to answer:
1. What is the name of the executable attached to the PTY? (What shells are popular? Should we focus our testing on them? Are there any common ones we are blind to that we should know about?)
- "Microsoft.Windows.Terminal.Connection" {e912fe7b-eeb6-52a5-c628-abe388e5f792}
- "ConPtyConnected" event
- "SessionGuid" value = WT_SESSION 
- "Client" value = Name of EXE
2. Is Acrylic used on a tab? And with what opacity? (Do people really want acrylic? Should it be default? What opacity is most pleasing in our context?)
- "Microsoft.Windows.Terminal.App" {24a1622f-7da7-5c77-3303-d850bd1ab2ed}
- "TabInformation" event
- "EventVer" value is now 1u
- "UseAcrylic" value is now TRUE/FALSE on the setting choice
- "TintOpacity" value is now Float on the setting choice
3. What font are people choosing? (Do people move away from Cascadia Code? Which ones are the most popular for us to validate when updating the renderer?)
- "Microsoft.Windows.Terminal.App" {24a1622f-7da7-5c77-3303-d850bd1ab2ed}
- "TabInformation" event
- "FontFace" value is now string font from settings
4. What keybindings do people choose to customize (Add or Remove)? (Are there extremely common keys that folks bind or unbind that we should have adjusted by default in a fresh install?)
- "Microsoft.Windows.Terminal.App" {24a1622f-7da7-5c77-3303-d850bd1ab2ed}
- "CustomKeybindings" event
- "Keybindings" value is the entire JSON segment that describes the user keybindings from `settings.json`.
5. Do people change their default profile from the PowerShell one we set? If so, to what? (Should we not set PowerShell as the default? Should we adjust the ranking of our dynamic generators to favor the most popular ones to bubble to the top?)
- "Microsoft.Windows.Terminal.App" {24a1622f-7da7-5c77-3303-d850bd1ab2ed}
- "CustomDefaultProfile" event
- "DefaultProfile" value is the GUID of the chosen profile

## Validation Steps Performed
1. Implemented the events
2. Launched the ETL channel viewer
3. Triggered the events
4. Saw the data come out

### CODE REVIEW

**Comment by miniksa** [comment]
Outstanding questions:

For 4: Since we can have people put arbitrary arguments into keybindings now, should we strip all that out of the JSON and reserialize it, change the format of our collection, or something like that to avoid collecting personally identifiable information?

**Comment by miniksa** [comment]
> Outstanding questions:
> 
> For 4: Since we can have people put arbitrary arguments into keybindings now, should we strip all that out of the JSON and reserialize it, change the format of our collection, or something like that to avoid collecting personally identifiable information?

I ran it through the serializer/deserializer to discard anything that isn't something we comprehend (comments and other such stuff.)

File: src/cascadia/TerminalApp/CascadiaSettingsSerialization.cpp [hunk 1]
```diff
@@ -129,6 +134,54 @@ std::unique_ptr<CascadiaSettings> CascadiaSettings::LoadAll()
...
+                TraceLoggingDescription("Event emitted when user has chosen a different default profile than hardcoded one on load/reload"),
+                TraceLoggingGuid(guid, "DefaultProfile", "ID of user-chosen default profile"),
+                TraceLoggingKeyword(MICROSOFT_KEYWORD_MEASURES),
+                TelemetryPrivacyDataTag(PDT_ProductAndServiceUsage));
+        }
+
+        // If the user had keybinding settings preferences, we want to learn from them to make better defaults
+        auto userKeybindings = resultPtr->_userSettings[JsonKey(KeybindingsKey)];
+        if (!userKeybindings.empty())
+        {
```
Comment (DHowett-MSFT) [inline src/cascadia/TerminalApp/CascadiaSettingsSerialization.cpp]: Hate to do this to you; feel free to shout at me if it's too much work.

Moving this into a function that's called _inside_ the `TraceLoggingWrite` macro will defer all the work, possibly _forever_ if tracing is not actively enabled for this session/user/machine.

This just seems costly to do every time. Thoughts?

File: src/cascadia/TerminalConnection/ConptyConnection.h [hunk 2]
```diff
@@ -41,6 +41,7 @@ namespace winrt::Microsoft::Terminal::TerminalConnection::implementation
         hstring _startingDirectory;
         hstring _startingTitle;
         guid _guid{}; // A unique session identifier for connected client
+        hstring _clientName{}; // The name of the process hosted by this ConPTY connection (as of launch).
```
Comment (DHowett-MSFT) [inline src/cascadia/TerminalConnection/ConptyConnection.h]: keeping in case we need it? or can it just be confined to a local

Comment (miniksa) [reply src/cascadia/TerminalConnection/ConptyConnection.h]: I thought we verbally mentioned that this information might be useful under the debugger so I thought "ehhh what the heck". I can move it back to a local if you want.

Comment (DHowett-MSFT) [reply src/cascadia/TerminalConnection/ConptyConnection.h]: Fair enough

File: src/cascadia/TerminalApp/CascadiaSettingsSerialization.cpp (hunk 1 above)
Comment (miniksa) [reply src/cascadia/TerminalApp/CascadiaSettingsSerialization.cpp]: Nnnnnn as far as I remember, that would count as "doing something cute" with the TraceLoggingMacros. But as far as I remember, there's something in the `conhost.exe` code that lets you check if a session is tracing or not so you can disable yourself from hard-processing work. Let me go see if I can get that here.

Comment (miniksa) [reply src/cascadia/TerminalApp/CascadiaSettingsSerialization.cpp]: OK, I found it and put it in.

File: src/cascadia/TerminalApp/lib/TerminalAppLib.vcxproj [hunk 3]
```diff
@@ -271,7 +271,7 @@
   <ItemDefinitionGroup>
     <ClCompile>
       <PrecompiledHeaderFile>pch.h</PrecompiledHeaderFile>
-      <AdditionalIncludeDirectories>..;$(OpenConsoleDir)\dep;$(OpenConsoleDir)\dep\jsoncpp\json;%(AdditionalIncludeDirectories);</AdditionalIncludeDirectories>
+      <AdditionalIncludeDirectories>..;$(OpenConsoleDir)\dep\jsoncpp\json;%(AdditionalIncludeDirectories);</AdditionalIncludeDirectories>
```
Comment (amit-msft) [inline src/cascadia/TerminalApp/lib/TerminalAppLib.vcxproj]: CLI11.hpp won't be accessible to TerminalAppLib pch.h that this change added https://github.com/microsoft/terminal/commit/830c22b73e3299cde552208597ddebffbeb0a898

Comment (DHowett-MSFT) [reply src/cascadia/TerminalApp/lib/TerminalAppLib.vcxproj]: `dep` was always on the include line, so adding it _earlier_ than `AdditionalIncludeDirectories` was simply done in error 😄

File: src/cascadia/TerminalApp/TerminalPage.cpp [hunk 4]
```diff
@@ -480,9 +480,13 @@ namespace winrt::TerminalApp::implementation
             g_hTerminalAppProvider, // handle to TerminalApp tracelogging provider
             "TabInformation",
             TraceLoggingDescription("Event emitted upon new tab creation in TerminalApp"),
+            TraceLoggingUInt32(1u, "EventVer", "Version of this event"),
             TraceLoggingUInt32(tabCount, "TabCount", "Count of tabs currently opened in TerminalApp"),
             TraceLoggingBool(usedManualProfile, "ProfileSpecified", "Whether the new tab specified a profile explicitly"),
             TraceLoggingGuid(profileGuid, "ProfileGuid", "The GUID of the profile spawned in the new tab"),
+            TraceLoggingBool(settings.UseAcrylic(), "UseAcrylic", "The acrylic preference from the settings"),
+            TraceLoggingFloat64(settings.TintOpacity(), "TintOpacity", "Opacity preference from the settings"),
+            TraceLoggingWideString(settings.FontFace().c_str(), "FontFace", "Font face chosen in the settings"),
```
[the remark this answers is not in the document]
Comment (amit-msft) [reply src/cascadia/TerminalApp/TerminalPage.cpp]: Doh! Didn't realize till now I had to "submit review" in order for my comments to show up for everyone 😐. A bit late to address these on this PR eh

File: src/cascadia/TerminalApp/lib/TerminalAppLib.vcxproj (hunk 3 above)
Comment (amit-msft) [reply src/cascadia/TerminalApp/lib/TerminalAppLib.vcxproj]: Ah 🙂

File: src/cascadia/TerminalApp/TerminalPage.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (miniksa) [reply src/cascadia/TerminalApp/TerminalPage.cpp]: I asked @zadjii-msft and he said this method on "page" is called for both creating a tab and creating a pane. So it should be fine for both scenarios.

**Pull request merged**